cluster/images/etcd/migrate-if-needed.sh

```diff
@@ -39,240 +39,43 @@
-#    attachlease commands.
-SKIP_STEP=true
-for step in ${SUPPORTED_VERSIONS}; do
-  if [ "${step}" = "${CURRENT_VERSION}" ]; then
-    SKIP_STEP=false
-  elif [ "${SKIP_STEP}" != "true" ]; then
-    # Do the migration step, by just starting etcd in this version.
-    START_VERSION="${step}"
-    START_STORAGE="${CURRENT_STORAGE}"
-    if ! start_etcd; then
-      # Migration failed.
-      echo "Starting etcd ${step} failed"
-      exit 1
-    fi
-    # Kill etcd and wait until this is down.
-    stop_etcd
-    CURRENT_VERSION=${step}
-    echo "${CURRENT_VERSION}/${CURRENT_STORAGE}" > "${DATA_DIRECTORY}/${VERSION_FILE}"
-  fi
-  if [ "$(echo ${CURRENT_VERSION} | cut -c1-2)" = "3." -a "${CURRENT_VERSION}" = "${step}" -a "${CURRENT_STORAGE}" = "etcd2" -a "${TARGET_STORAGE}" = "etcd3" ]; then
-    # If it is the first 3.x release in the list and we are migrating
-    # also from 'etcd2' to 'etcd3', do the migration now.
-    echo "Performing etcd2 -> etcd3 migration"
-    START_VERSION="${step}"
-    START_STORAGE="etcd3"
-    ETCDCTL_CMD="${ETCDCTL:-/usr/local/bin/etcdctl-${START_VERSION}}"
-    ETCDCTL_API=3 ${ETCDCTL_CMD} migrate --data-dir=${DATA_DIRECTORY}
-    echo "Attaching leases to TTL entries"
-    # Now attach lease to all keys.
-    # To do it, we temporarily start etcd on a random port (so that
-    # apiserver actually cannot access it).
-    if ! start_etcd; then
-      echo "Starting etcd ${step} in v3 mode failed"
-      exit 1
-    fi
-    # Create a lease and attach all keys to it.
-    ${ATTACHLEASE} \
-      --etcd-address http://127.0.0.1:${ETCD_PORT} \
-      --ttl-keys-prefix "${TTL_KEYS_DIRECTORY:-${ETCD_DATA_PREFIX}/events}" \
-      --lease-duration 1h
-    # Kill etcd and wait until this is down.
-    stop_etcd
-    CURRENT_STORAGE="etcd3"
-    echo "${CURRENT_VERSION}/${CURRENT_STORAGE}" > "${DATA_DIRECTORY}/${VERSION_FILE}"
-  fi
-  if [ "$(echo ${CURRENT_VERSION} | cut -c1-4)" = "3.1." -a "${CURRENT_VERSION}" = "${step}" -a "${CURRENT_STORAGE}" = "etcd3" ]; then
-    # If we are upgrading to 3.1.* release, if the cluster was migrated
-    # from v2 version, the v2 data may still be around. So now is the
-    # time to actually remove them.
-    echo "Remove stale v2 data"
-    START_VERSION="${step}"
-    START_STORAGE="etcd3"
-    ETCDCTL_CMD="${ETCDCTL:-/usr/local/bin/etcdctl-${START_VERSION}}"
-    if ! start_etcd; then
-      echo "Starting etcd ${step} in v3 mode failed"
-      exit 1
-    fi
-    ${ETCDCTL_CMD} --endpoints "http://127.0.0.1:${ETCD_PORT}" rm --recursive "${ETCD_DATA_PREFIX}"
-    # Kill etcd and wait until this is down.
-    stop_etcd
-    echo "Successfully remove v2 data"
-    # Also remove backup from v2->v3 migration.
-    rm -rf "${BACKUP_DIR}"
-  fi
-  if [ "${CURRENT_VERSION}" = "${TARGET_VERSION}" -a "${CURRENT_STORAGE}" = "${TARGET_STORAGE}" ]; then
-    break
-  fi
-done
-
-echo "$(date +'%Y-%m-%d %H:%M:%S') Migration finished"
+/usr/local/bin/migrate \
```

Comment:
1. Did you miss binDir in the list? (line 49 https://github.com/kubernetes/kubernetes/pull/61241/files#diff-755f39aaeeec98fb826929445011dde0)
2. Would it be better if they are in the same order as in migrateOpts struct in migrate.go? (minor)

Reply:
#1 - The default value is correct, we don't need to explicitly set it.
#2 - I don't feel strongly about it.  The flags are also in a different order in the `migrate -h` output.  We did keep the relative order of all flags and env vars throughout `migrate-if-needed.sh`, which seems like the most important place to be consistent.

Reply:
Sounds good.